python/tutorials/gluon/08-warp-specialization.py

```diff
@@ -0,0 +1,655 @@
+            mbarrier.wait(load_ready_bars.index(load_state.index), load_state.phase)
+            tcgen05_mma(p.a_bufs.index(load_state.index), p.b_bufs.index(load_state.index), acc_buf, use_acc=use_acc)
+            tcgen05_commit(load_empty_bars.index(load_state.index))
+            load_state = load_state.next()
+            use_acc = True
+        # Complete the accumulator asynchronously.
+        tcgen05_commit(acc_ready_bars.index(acc_state.index))
+        acc_state = acc_state.next()
+
+
+# Helper for splitting a tensor along N. This only works for certain layouts.
+# For our kernel, this only works for BLOCK_M=128 and num_warps=4.
+@gluon.jit
+def _split_n(x, SUBTILE_FACTOR: gl.constexpr):
+    split_count: gl.constexpr = SUBTILE_FACTOR.bit_length() - 1  # log2
+    xs = (x, )
+    for _ in gl.static_range(split_count):
+        next_xs = ()
+        for j in gl.static_range(len(xs)):
+            x = xs[j]
+            # Reshape to (M, 2, N//2) then permute so that tensor elements
+            # remain contiguous along N.
+            next_xs += x.reshape(x.shape[0], 2, x.shape[1] // 2).permute(0, 2, 1).split()
+        xs = next_xs
+    return xs
+
+
+@gluon.jit
+def matmul_epilogue_partition(p, SchedulerImpl: gl.constexpr):
+    BLOCK_M: gl.constexpr = p.c_desc.block_type.shape[0]
+    BLOCK_N: gl.constexpr = p.c_desc.block_type.shape[1]
+    dtype: gl.constexpr = p.c_desc.dtype
+
+    acc_empty_bars = p.acc_empty_bars
+    acc_ready_bars = p.acc_ready_bars
+    acc_state = Counter.create(0, p.acc_empty_bars.shape[0])
+    acc_layout: gl.constexpr = get_tmem_32x32b_reg_layout(BLOCK_M, BLOCK_N, [BLOCK_M, BLOCK_N], p.num_warps)
+    SPLIT_N: gl.constexpr = BLOCK_N // p.SUBTILE_FACTOR
+    acc_smem = gl.allocate_shared_memory(dtype, [BLOCK_M, SPLIT_N], p.c_desc.layout)
+
+    scheduler = SchedulerImpl.initialize(p.c_desc.shape[0], p.c_desc.shape[1], BLOCK_M, BLOCK_N)
+    for idx in range(scheduler.get_num_tiles()):
+        pid_m, pid_n = scheduler.get_tile(idx)
+        off_m = pid_m * BLOCK_M
+        off_n = pid_n * BLOCK_N
+
+        # Wait for the accumulator. Since BLOCK_N=256, we need to interleave
+        # the TMEM loads with the SMEM stores to avoid spilling.
+        mbarrier.wait(acc_ready_bars.index(acc_state.index), acc_state.phase)
+        acc = p.acc_bufs.index(acc_state.index).load(acc_layout)
+        acc_state = acc_state.next()
+
+        accs = _split_n(acc, p.SUBTILE_FACTOR)
+        for i in gl.static_range(len(accs)):
+            acc = accs[i].to(dtype)
+            tma.store_wait(pendings=0)  # overlap with downcast
+            acc_smem.store(acc.to(dtype))
+            # Arrive after the first SMEM store and rely on ptxas to interleave.
+            if i == 0:
+                mbarrier.arrive(acc_empty_bars.index(acc_state.index), count=1)
+            fence_async_shared()
+            tma.async_copy_shared_to_global(p.c_desc, [off_m, off_n + SPLIT_N * i], acc_smem)
+        tma.store_wait(pendings=0)
```

Comment:
?
```suggestion
    tma.store_wait(pendings=0)
```

Reply:
The original placement was intentional but you're right that we can move it here.